cpp/src/arrow/array/data.cc

```diff
@@ -151,6 +151,13 @@ void ArraySpan::SetMembers(const ArrayData& data) {
     }
   }
 
+  Type::type type_id = this->type->id();
+  if (data.buffers[0] == nullptr && type_id != Type::NA &&
+      type_id != Type::SPARSE_UNION && type_id != Type::DENSE_UNION) {
```

Comment:
nit: there's [`arrow::internal::HasValidityBitmap` in `arrow/type.h`](https://github.com/apache/arrow/blob/d49d8de3a9158f66ff7073fe077c5c292e38de6e/cpp/src/arrow/type.h#L2043-L2052)